proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,483 @@
+  ::
+
+    let
+      name1 = _ :: Text
+      age1 = _ :: Int
+      name2 = _ :: Text
+      age2 = _ :: Int
+
+      textExample1 = name1 <> " (age: " <> T.pack (show age1) <> ") encountered " <> name2 <> " (age: " <> T.pack (show age2) <> ")"
+
+      textExample2 = T.pack $ T.unpack name1 <> " (age: " <> show age1 <> ") encountered " <> T.unpack name2 <> " (age: " <> show age2 <> ")"
+
+      textExample3 = T.unwords
+        [ name1
+        , "(age: " <> T.pack (show age1) <> ")"
+        , "encountered"
+        , name2
+        , "(age: " <> T.pack (show age2) <> ")"
+        ]
+
+* ``printf`` is partial and unsafe, which especially safety-conscious people might always stay away from anyway. Using a safer ``printf`` like ``formatting`` induces a third-party dependency, which is admittedly lightweight, but isn't as seamless of an integration as native string interpolation would be
+
+* Quasiquotes induces a dependency on Template Haskell, which a lot of people avoid out of principle. Most QuasiQuoters also add a dependency on ``haskell-src-exts`` to parse arbitrary Haskell expressions, which could technically be avoided by using something like ``ghc-meta`` (`repo <https://github.com/noughtmare/ghc-meta>`_, `GHC issue <https://gitlab.haskell.org/ghc/ghc/-/issues/20862>`_), but this isn't in wide use yet.
+
+If Haskell had native string interpolation, it would have the benefit and safety of the current third-party quasiquotes without the need for Template Haskell, and be able to take advantage of features like `multiline strings <https://github.com/ghc-proposals/ghc-proposals/pull/569>`_.
+::
+
+  error s"Expected: ${x + y}, got: ${result}"
+
+  let textExample = s"${name1} (age: ${age1}) encountered ${name2} (age: ${age2})"
+
+Proposed Change Specification
+-----------------------------
+
+Grammar
+~~~~~~~
+
+Add a new ``-XStringInterpolation`` extension, which, when enabled, adds a new expression to `Chapter 3 of the Haskell 2010 Report <https://www.haskell.org/onlinereport/haskell2010/haskellch3.html#x8-220003>`_ after do-expressions in the ``exp`` rule:
+
+.. code-block:: abnf
+
+  exp → ...
+      | case exp of { alts }                    (case expression)
+      | do { stmts }                            (do expression)
+      | s"   interpolatedString "               (interpolated strings)
+      | s""" interpolatedStringMultiline """    (interpolated multiline strings)
+      | fexp
+
+Note that no space is allowed between the ``s`` and ``"``.
+
+The ``interpolatedString`` and ``interpolatedMultilineString`` rules are new rules defined as follows:
+
+.. code-block:: abnf
+
+  interpolatedString               → {graphic⟨\ | "   | ${⟩ | space | escape | gap} [ ${ exp } interpolatedString ]
+  interpolatedStringMultiline      → {{whitechar} interpolatedStringMultilineLine}
+  interpolatedStringMultilineLine  → {graphic⟨\ | """ | ${⟩ | space | escape | gap} [ ${ exp } interpolatedStringMultilineLine ]
```

Comment:
Here it looks like you want to specify the lexical structure. This should come before the syntactic specification for `exp`!
Furthermore, you cannot mention the syntactic non-terminal `exp` while specifying lexical structure.

Also you do not extend `charesc` with `$` as I recommended in https://github.com/ghc-proposals/ghc-proposals/pull/570#issuecomment-2270722558. Otherwise, how would escape `$`?

---

Here is one idea for a specification. First, lexical structure:

```
interpolatedString → 's' " {graphic⟨'\' | '"'   | '$' '{'⟩ | space | escape | gap} [ '$' '{' { any⟨'}' | '"'⟩ } '}' ] } '"'
```

a post-processing step between lexer and parser must then split the single `interpolatedString` token along the the `{ any⟨'}' | '"'⟩ }` parts, which correspond to where the parser needs to recognise `exp`.

We will have 3 kinds of new tokens:

1. Every interpolated string gets born by the lexer as an `InterpolatedStringBalanced`. If there are no `${ ... }` in the string, it stays that way.
2. After lexing, the `InterpolatedStringBalanced` gets *segmented*: i.e. broken apart after each `${` and before each `}`. We get one token kind `InterpolatedStringBegin` that includes `${`, and another kind of token `InterpolatedStringEnd` which includes `}`.
3. Finally, we *re-lex* the `any` part that will constitute the expression.

Example:

```
Initially:
   s"fooooooooooo ${baaaaaaaaaar} baaaaaaaaaz ${qoooooooooooo} quuuuuuuz"
   <------------------------------ISClosed------------------------------> 
After segmentation and relexing:
   s"fooooooooooo ${baaaaaaaaaar} baaaaaaaaaz ${qoooooooooooo} quuuuuuuz"
   <--ISBegin------><--varid---><--ISClosed----><---varid--->-<--ISEnd--> 
``` 

Now we add the following production rules to the parser:

```
exp -> interpolatedStringClosed
exp -> interpolatedStringBegin exp { interpolatedStringClosed exp } interpolatedStringEnd
```

Something similar should work for multiline strings (do we have a specification for their lexical syntax?).

Reply:
ahh your comment just made the difference between Lexer.x and Parser.y click for me. I wasn't really sure what the difference was 😅 I was also unsure if it was possible to re-lex in the parser; how would that work in GHC? Or are you suggesting the re-lex might happen in a later step, like renaming?

Yeah, the specification for multiline strings is here: https://github.com/brandonchinn178/ghc-proposals/blob/master/proposals/0569-multiline-strings.rst#bnf

I pushed a new commit with some changes. Probably still needs work, but let me know if it's trending in the right direction